agent/consul/state/acl.go

```diff
@@ -447,8 +496,13 @@ func (s *Store) aclTokenGetTxn(tx *memdb.Txn, ws memdb.WatchSet, value, index st
 
 	if rawToken != nil {
 		token := rawToken.(*structs.ACLToken)
```

Comment:
In my PR chain to work around the same problem quickly, I'd resorted to changing lines like this to `token := rawToken.(*structs.ACLToken).Clone()` and leaving the rest of it alone. The upside was the overall code delta to avoid the badness was very light, with the downside of there being a slightly higher short-lived alloc memory pressure.

Looking at this COW version, I can't help but feel like this solution is possibly a bit too complicated/sprawly and that maybe the dumber clone-everything solution would be practically better even with the alloc expense.

Reply:
So the simple solution of just COW it all the time has much less code sprawl. My initial implementation was just that. However for replication purposes I was thinking it would be good to provide a way to request the raw data without modification. This actually would perform even better than the initial buggy implementation. It does however mean that everywhere that requests tokens must indicate whether they want that link fixup done.

I certainly could back out all those other changes and we could just take the hit every time we lookup a token with stale policy links which will probably degrade into every token resolution over time.

Reply:
I wonder what the alloc and timing diff is between both variations for a "reasonable" dataset and request count.